User can view all items up for auction.

xray opened the pull request.

xray committed:
add tests for auction controller

xray committed:
add function to get list of auctions

xray committed:
add page to view auctions

xray committed:
link auctions from home page

kevbuchanan commented on test/renaissance_web/controllers/login_controller_test.exs:
```diff
@@ -1,17 +1,19 @@
     conn =
-      post(build_conn(), "/login", %{"user" => %{email: "mail@mail.com", password: "password"}})
+      build_conn()
+      |> post("/register", %{"user" => @user_params})
```
Rather than relying on using the controllers to put the connection in the expected state (signed in), user a setup block to just create a user and assign to the session, returning a signed in connection.

kevbuchanan commented on lib/renaissance_web/controllers/auction_controller.ex:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  def create(conn, params) do
+    changeset =
```
This isn't a changeset.

rstein66 replied:
Resolved with 159a775e493da7136d042a4f76b690b4d8cb2aea

kevbuchanan commented on lib/renaissance/auctions/auctions.ex:
```diff
@@ -0,0 +1,66 @@
+    Repo.get_by(Auction, title: title)
+  end
+
+  def get_all_auctions() do
```
Rather than returning a list of an unspecified map here, I'd expect this to return `Auction` struct, with the seller preloaded:

```
Auction
|> preload(:user)
|> Repo.all()
```

This also leads to the mismatch between what you seem to want to call the `:user` association (`:seller`) and what you've called it on the model. In your ecto schema you could change the `:user` association to `:seller`:

`belongs_to :seller, User`

Then in User:

`has_many :auctions, Auction, foreign_key: :seller_id`

kevbuchanan commented on .gitignore:
```diff
@@ -175,3 +175,73 @@ modules.xml
 
 # End of https://www.gitignore.io/api/phoenix
+
+# Created by https://www.gitignore.io/api/macos
```
Same comment as the other PR (as well as a bunch of others). This file is getting out of hand.

kevbuchanan commented on lib/renaissance/auctions/auction.ex:
```diff
@@ -18,8 +19,8 @@ defmodule Renaissance.Auction do
     user
-    |> cast(attrs, @required_fields ++ @optional_fields)
-    |> validate_required(@required_fields)
+    |> cast(attrs, [:title, :description, :user_id, :price, :end_date])
```
Why this change? Using the `@required_fields` and `@optional_fields` is the preferred way here. Cuts down on duplication and makes it easier to add new fields.

xray replied:
This was an issue with rebasing from [feature-create-auction](https://github.com/xray/Renaissance/tree/feature-create-auction), this will be fixed.

xray commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[auctions] User can view all items up for auction.](https://trello.com/c/xOCbcYI5/5-auctions-user-can-view-all-items-up-for-auction)

rstein66 committed:
Remove editorconfig file

rstein66 committed:
Ensure core fields [in migration] are not null

rstein66 committed:
Improve adherence to Phoenix route name convention

Reference:
https://hexdocs.pm/phoenix/routing.html
https://hexdocs.pm/phoenix/controllers.html#actions

ssunday commented on lib/renaissance_web/router.ex:
```diff
@@ -16,14 +16,15 @@ defmodule RenaissanceWeb.Router do
     post "/login", LoginController, :verify
-    get "/auctions", AuctionController, :list
+
+    get "/auctions", AuctionController, :index
```
You can use resources for this now, right

ssunday commented on priv/repo/migrations/20190404190924_create_users.exs:
```diff
@@ -3,8 +3,8 @@ defmodule Renaissance.Repo.Migrations.CreateUsers do
     create table(:users) do
-      add :email, :string
-      add :password_hash, :string
+      add :email, :string, null: false
```
This migration is already checked in, right? You shouldn't modify old migrations instead make new ones. If this had already been deployed, it wouldn't actually take effect

ssunday replied:
Also it doesn't convey the history of the changes

rstein66 replied:
Too true–will resist this cheat going forward :flushed:

ssunday commented on test/renaissance_web/controllers/auction_controller_test.exs:
```diff
@@ -20,81 +36,65 @@ defmodule RenaissanceWeb.AuctionControllerTest do
   end
 
-  test "POST /auctions/new fails to create auction when title is blank" do
+  test "POST /auctions fails to create auction when title is blank" do
```
Couldn't you do `%{@aucton_one | title: ""} or however to merge the maps?

rstein66 replied:
Resolved with 159a775e493da7136d042a4f76b690b4d8cb2aea

rstein66 committed:
Improve readability of CSS

rstein66 committed:
Reduce test run time

rstein66 committed:
Remove unused deps

rstein66 committed:
Simplify auction form input handling & testing

ssunday commented on assets/css/app.css:
```diff
@@ -175,7 +183,7 @@ body {
 .auction-detail {
-    margin: 0px;
+    margin: 0;
     font-family: Roboto, sans-serif;
```
This style of saying roboto is diff than above

ssunday commented on lib/renaissance/auctions/auction.ex:
```diff
@@ -26,24 +26,22 @@ defmodule Renaissance.Auction do
+    {_, amount} = Money.Ecto.Amount.Type.dump(get_change(changeset, :price, 0))
 
-    if Money.compare(item_price || one_dollar, zero_dollars) == 1 do
+    if 0 < amount do
```
could you reverse this ? `0 < amount` is harder to understand than `amount > 0` IMO

rstein66 replied:
Sure, will do.
_(I find it more intuitive to have the smaller value on the left, but don't feel strongly.)_

elizabethengelman commented on lib/renaissance/auctions/auctions.ex:
```diff
@@ -4,9 +4,15 @@ defmodule Renaissance.Auctions do
+    details =
+      details
       |> add_user_id(email)
-      |> format_end_date()
```
Why are we removing the `format_end_date` step here?

rstein66 replied:
The form is now using a Phoenix date time input method and so it is able to understand the input without needing to be told

elizabethengelman commented on test/renaissance/auctions/auction_test.exs:
```diff
@@ -4,136 +4,78 @@ defmodule Renaissance.Test.AuctionTest do
+  @now DateTime.utc_now()
+  @after_one_day %{@now | day: @now.day + 1}
+
+  @valid_params %{
```
I like this refactoring to reduce duplication.

rstein66 committed:
Fix & test db associations; clean up

Clean up includes:
- Renaming db column for consistency & clarity
- Renaming methods to adhere more closely with
routing/action naming conventions
(e.g. `:register` -> `:create`)
- Utilize built-in methods
(e.g., custom `get_all_auctions` query -> `Repo.all`)
- Reducing test code duplication

---

_n.b.,_ unintentionally large commit;
decided not to split up changes after the fact
because of their interdependence, but
will aim to be more proactive going forward
